solidity/ecdsa/contracts/WalletFactory.sol

```diff
@@ -0,0 +1,270 @@
+    }
+
+    /// @notice Registers the caller in the sortition pool.
+    function registerOperator() external {
+        address operator = msg.sender;
+
+        require(
+            !sortitionPool.isOperatorInPool(operator),
+            "Operator is already registered"
+        );
+
+        sortitionPool.insertOperator(
+            operator,
+            staking.eligibleStake(operator, address(this))
+        );
+    }
+
+    /// @notice Updates the sortition pool status of the caller.
+    function updateOperatorStatus() external {
+        sortitionPool.updateOperatorStatus(
+            msg.sender,
+            staking.eligibleStake(msg.sender, address(this))
+        );
+    }
+
+    // TODO: Revisit to implement mechanism for a fresh wallet creation.
+    function requestNewWallet() external onlyOwner {
+        dkg.lockState();
+        dkg.start(
+            uint256(keccak256(abi.encodePacked(relayEntry, block.number)))
+        );
+    }
+
+    function submitDkgResult(DKG.Result calldata dkgResult) external {
+        dkg.submitResult(dkgResult);
+
+        address clonedWalletAddress = createClone(address(masterWallet));
+        require(
+            clonedWalletAddress != address(0),
+            "Cloned wallet address is 0"
+        );
+
+        Wallet wallet = Wallet(clonedWalletAddress);
+
+        wallets.push(wallet);
+
+        wallet.init(
+            address(this),
+            hashGroupMembers(
+                dkgResult.members,
+                dkgResult.misbehavedMembersIndices
+            ),
+            keccak256(dkgResult.groupPubKey)
+        );
+
+        emit WalletCreated(address(wallet), keccak256(abi.encode(dkgResult)));
+    }
+
+    /// @notice Notifies about DKG timeout. Pays the sortition pool unlocking
+    ///         reward to the notifier.
+    function notifyDkgTimeout() external {
+        dkg.notifyTimeout();
+
+        // TODO: Implement transferDkgRewards
+        // transferDkgRewards(msg.sender, sortitionPoolUnlockingReward);
+
+        dkg.complete();
+    }
+
+    function approveDkgResult(DKG.Result calldata dkgResult) external {
+        uint32[] memory misbehavedMembers = dkg.approveResult(dkgResult);
+
+        Wallet latestWallet = wallets[wallets.length - 1];
+        latestWallet.activate();
+
+        // TODO: Transfer Wallet's ownership to WalletManager
+
+        // TODO: Transfer DKG rewards and disable rewards for misbehavedMembers.
+        misbehavedMembers;
+
+        emit WalletActivated(address(latestWallet));
+
+        dkg.complete();
+    }
+
+    function challengeDkgResult(DKG.Result calldata dkgResult) external {
+        Wallet latestWallet = wallets[wallets.length - 1];
+        require(
+            latestWallet.activationBlockNumber() == 0,
+            "The latest registered wallet was already activated"
+        );
+
+        (bytes32 maliciousResultHash, uint32 maliciousSubmitter) = dkg
+            .challengeResult(dkgResult);
+
+        wallets.pop();
```

Comment:
Don't we need to destroy the wallet contract as well? `selfdestruct(address(wallet))`;

And another question if we delete wallets that were valid but its age / active period has passed?

Reply:
> Don't we need to destroy the wallet contract as well? selfdestruct(address(wallet));

I'm not sure we need to do this. The destruct would have to be called from the `Wallet` contract, and I'm not sure if it will work with a clone/proxy patter properly. `selfdestruct` transfers ETH from the contract so we would need to make sure we handle such a scenario.

> And another question if we delete wallets that were valid but its age / active period has passed?

I guess we won't delete them but change their state.

Reply:
ok, I thought that we can save some gas on selfdestruct, but after some testing it doesn't look like we save anything. So let's leave it as is.